sunxspex/sun_xspec.py

```diff
@@ -0,0 +1,349 @@
+
+These models can be added via pyxspec:
+
+thick=sun_xspec.ThickTargetModel()
+xspec.AllModels.addPyMod(thick.model, thick.ParInfo, 'add')
+
+or XSPEC: (tbd but read the documentation)
+"""
+import numpy as np
+import pandas as pd
+#from scipy.special import lpmv
+from sunxspex.emission import split_and_integrate, split_and_integrate0 #, BrokenPowerLawElectronDistribution, bremsstrahlung_cross_section, collisional_loss
+from sunxspex import thermal
+from sunxspex import constants
+from astropy import units as u
+#import astropy.constants as c
+#from quadpy.c1 import gauss_legendre
+from datetime import datetime as dt
+import logging
+
+logging.basicConfig(filename='xspec.log', level=logging.DEBUG) #for now this helps catch Python errors that are not printed to the terminal by XSPEC
+
+# Central constant management
+global const #whyyyyyy there has got to be a better way to do this
+const = constants.Constants()
+
+class XspecModel:
+    def __init__(self):
+        self.ParInfo=''
+        self.model=None
+        self.description=None
+        #should I deal with constants here?
+
+    def __repr__(self):
+        return self.description
+
+    def print_ParInfo(self):
+        '''print parameter info in a readable format with headers '''
+        headers="Parameter,Unit,Default,Hard Min,Soft Min,Soft Max,Hard Max,Delta".split(',')
+        partable=np.array([[p.split('  ') for p in self.ParInfo]]).reshape((len(self.ParInfo),8)).T
+        df=pd.DataFrame({h:r for h,r in zip(headers,partable)}, index=range(1,len(self.ParInfo)+1))
+        return df #prints nicely like this...
+
+    def other_method(self):
+        '''such as: easily set parameter defaults from a tuple or dictionary, descriptions of parameters, etc'''
+        raise NotImplementedError
+
+
+class ThickTargetModel0(XspecModel): #original integration for comparison
+    def __init__(self):
+        self.ParInfo=(
+        "a0  1e-35  100.0  1.0  10.0  1e6  1e7  1.0",
+        "p  \"\"  4.0  1.1  1.5  15.0  20.0  0.1",
+           "eebrk  keV  150.0  1.0  5.0  100.  1e5  0.5" ,
+         "q  \"\"  6.0  0.0  1.5  15.0  20.0  0.1",
+         "eelow  keV  20.0  0.0  1.0  100.  1e3  1.0" ,
+         "eehigh  keV  3200.0  1.0  10.0  1e6  1e7  1.0"
+        ) #default parameters from OSPEX
+        self.model=self.thick2_original
+        self.description=f"Thick-target bremsstrahlung '{self.model.__name__}'"
+
+    @staticmethod
+    def thick2_original(photon_energies, params, flux):
+        """
+        tuples or lists containing photon energies, params
+              flux is empty list of length nE-1
+
+              The input array of energy bins gives the boundaries of the energy bins
+              and hence has one more entry than the output flux arrays.
+
+              The output flux array for an additive model should be in terms of photons/cm$^2$/s
+              (not photons/cm$^2$/s/keV) i.e. it is the model spectrum integrated over the energy bin.
+        """
+
```

Comment:
Why doesn't this method just call the `bremsstrahlung_thick_target` function instead of reimplementing it here?